crates/bevy_ecs/src/bundle.rs

```diff
@@ -0,0 +1,233 @@
+pub use bevy_ecs_macros::Bundle;
+
+use crate::{
+    component::{Component, ComponentFlags, ComponentId, Components, StorageType, TypeInfo},
+    entity::Entity,
+    storage::{SparseSetIndex, SparseSets, Table},
+};
+use bevy_ecs_macros::all_tuples;
+use std::{any::TypeId, collections::HashMap};
+
+/// A dynamically typed ordered collection of components
+///
+/// See [Bundle]
+pub trait DynamicBundle: Send + Sync + 'static {
+    /// Gets this [DynamicBundle]'s components type info, in the order of this bundle's Components
+    fn type_info(&self) -> Vec<TypeInfo>;
+
+    /// Calls `func` on each value, in the order of this bundle's Components. This will "mem::forget" the bundle
+    /// fields, so callers are responsible for dropping the fields if that is desirable.
+    fn put(self, func: impl FnMut(*mut u8));
+}
+/// A statically typed ordered collection of components
+///
+/// See [DynamicBundle]
+pub trait Bundle: DynamicBundle {
+    /// Gets this [Bundle]'s components type info, in the order of this bundle's Components
+    fn static_type_info() -> Vec<TypeInfo>;
```

Comment:
`type_info` and `static_type_info` always return the same value, right? If so `static_type_info` can be removed and instead `type_info` can be called.

Reply:
Oh, `static_type_info` doesn't take an `&self` param unlike `type_info`.